### PR TITLE
移行前プリフライトを実装する

### LINKED ISSUE #738: [Persistence v2] Cutover 前に migration preflight を実行し実ユーザーデータの矛盾・容量・identity collision を検出する
Parent: #724

Depends on: #711, #712, #725, #735, #736

Feeds into: #728, #729

Blocks: Phase 1 source-of-truth cutover

## 優先度

P0

## 背景

現行storageには段階的migrationの履歴があり、duplicate domain、mapping conflict、mixed `urls` / `urlIds`、schema違反record等が実データに存在し得る。

Persistence Model v2 migrationを初回cutover時にいきなり実行すると、初めてproduction data shapeの矛盾を知ることになる。

migration failure時のrecoveryを用意しても、cutover releaseで大量のユーザーが同時にfailureする設計は避けたい。

このIssueではsource-of-truth cutover前に**read-only migration preflight**を導入し、legacy sourceを変更せずにmigration可能性・data conflict・capacity riskを検出する。

## 目的

- migration release前にcurrent user dataの問題を検出する
- source storageを変更せずpreflightする
- typed conflict / capacity / URL identity collisionを把握する
- migration成功見込みを確認してから#729 cutoverへ進む
- repairが必要なuserをcutover前に識別する

## read-only invariant

preflightはsource dataへwriteしない。

禁止:

- existing forgiving getterによる自動repair
- schema違反recordのdrop + write-back
- missing timestampへの`Date.now()`補完 + source save
- hostname normalization結果のsource overwrite
- migration completed flag更新
- IndexedDB source-of-truth cutover

raw source readerからsnapshotを取得しpure analysisする。

## preflight pipeline

```text
raw chrome.storage snapshot read
  ↓
read failure / missing key classification
  ↓
legacy schema parse
  ↓
legacy source integrity pre-check
  ↓
Persistence v2 dry mapping
  ↓
URL identity collision analysis
  ↓
v2 integrity check
  ↓
capacity / target size estimate
  ↓
preflight report
```

IndexedDBへtemporary targetを書く必要がある場合もsource of truthへ切り替えず、isolated staging DB / migration IDで管理する。

## RawLegacyStorageReader

migration / preflight専用readerを定義する。

```ts
type RawLegacyStorageReader = {
  readonly readSnapshot: () => Promise<RawLegacyStorageSnapshot>
}
```

既存`getUrlRecords()` / `getCustomProjects()` / migration getter等をsource readerとして利用しない。

理由:

- read errorを`[]`へ潰す可能性
- invalid recordをdropする可能性
- timestamp補完する可能性
- source storageへrepair結果を書き戻す可能性

以下を必ず区別する。

```text
key missing
!=
empty array
!=
storage read failure
!=
invalid stored value
```

error例:

```ts
type MigrationSourceReadErrorCode =
  | 'MIGRATION_SOURCE_READ_FAILED'
  | 'MIGRATION_SOURCE_INVALID_TYPE'
  | 'MIGRATION_SOURCE_PARTIAL_READ'
```

## conflict detection

最低限:

- duplicate URL ID
- duplicate exact URL
- normalized URL collision
- nested `urls` / `urlIds` mismatch
- dangling URL reference
- duplicate canonical domain collection
- ParentCategory relation conflict
- domainCategoryMappings conflict
- CustomProject invalid schema
- notes / category metadata dangling reference
- project order conflict
- timestamp relation / missing timestamp quality

## URL identity collision report

#725で確定するURL identity semanticsを利用する。

例:

```text
legacy exact URL count: 10,000
v2 canonical URL count: 9,500
collision candidate: 500
```

500件をsilent mergeしない。

report:

```ts
type UrlIdentityCollisionSummary = {
  readonly collisionCount: number
  readonly collisionKinds: readonly UrlIdentityCollisionKind[]
}
```

raw URLをproduction logへ出さない。

自動merge可能 / requires review / migration blockingを分類する。

## capacity preflight

#735のpolicyに従い以下を可能な範囲で確認する。

- source approximate bytes
- expected v2 target bytes
- migration temporary duplication headroom
- storage usage / quota estimate

capacity riskが高い場合はcutoverを開始しない。

## preflight status

migration control stateとは分けて管理する。

例:

```ts
type MigrationPreflightStatus =
  | { readonly status: 'not-run' }
  | { readonly status: 'healthy'; readonly checkedAt: number }
  | {
      readonly status: 'blocked'
      readonly checkedAt: number
      readonly issueCodes: readonly MigrationPreflightIssueCode[]
    }
```

preflight healthyだけでmigration completedとは扱わない。

source fingerprintを持ち、preflight後sourceが変更された場合は結果を無効化する。

```text
preflight source fingerprint A
↓
normal write
↓
source fingerprint B
↓
preflight stale
```

## rollout options

以下を比較する。

### Option A: preflight releaseをmigration releaseより先行

```text
Release N
→ read-only preflight
→ conflict detection
→ repair / fixture追加

Release N+1
→ actual migration / cutover
```

最も安全。

### Option B: same releaseでpreflight → migration

preflight healthy userのみ続けてmigration。

blocking issueがあるuserはmigrationせずrecovery flowへ。

Parent #724では原則Option Aを優先し、schedule / release cadence上困難ならOption Bを明示的に判断する。

## privacy

preflight resultをexternal telemetryへ送らない。

local diagnosticで保持可能:

- entity counts
- issue codes
- collision counts
- approximate bytes
- timestamp quality summary

保持禁止:

- raw URL
- title
- notes
- AI prompt/history
- attachment content

## test

- raw read error != empty source
- missing key
- empty valid source
- invalid source type
- source mutationが発生しない
- duplicate URL / normalized collision
- relation conflict
- capacity blocked
- healthy report
- source fingerprint changed → preflight stale
- blocked userはcutoverしない

## 受け入れ条件

- [ ] dedicated raw legacy storage readerがある
- [ ] existing forgiving getterをpreflight / migration source readで使わない
- [ ] preflightがlegacy sourceを変更しない
- [ ] missing / empty / read failure / invalid typeを区別する
- [ ] URL identity collisionを検出する
- [ ] current legacy relation conflictをtypedに検出する
- [ ] capacity riskをpreflightできる
- [ ] source fingerprintによりstale preflightを無効化できる
- [ ] blocked userは#729 cutoverへ進まない
- [ ] safe diagnosticにraw user contentを含めない
- [ ] preflight先行release可否が決定している

## 注意点

- preflight healthyをmigration successとみなさない
- preflightのためにsourceをrepair / normalizeして書き戻さない
- conflict countを減らすためsilent drop / silent mergeしない
- production telemetryを新規追加しない

### PR DESCRIPTION
**Pull request by haruto-yamada1**
## 原因

現行の legacy storage reader は normalize-on-read / repair を含み、移行前の raw source を変更せずに検査する authority がありませんでした。また、#727 の bootstrap barrier には承認済み source fingerprint を読む seam はあるものの、preflight status の生成・stale invalidation・blocked user の recovery 導線が未実装でした。

## 採用した解決

- repairing getter を通らない get-only raw reader で migration source keys を一括取得し、missing / present-empty / read failure / invalid type を分離しました。
- raw snapshot を pure analyzer へ渡し、`exact-url-v1`、nested/canonical URL、parent/category/project relation、timestamp provenance、#712 integrity、#735 capacity を typed issue として fail-closed 判定します。
- #727 と同じ cross-context Web Lock を使い、exclusive snapshot → lock 外 analysis/capacity → exclusive re-read/save として TOCTOU を閉じました。
- versioned SHA-256 fingerprint と別 control-plane record `tabbin:migrationPreflight:v1` を追加し、source 変更時は `stale` にします。healthy/current match のみ #728 の既存 seam へ公開します。
- blocked/stale notice を app entrypoint に接続し、安全な集計診断コピー、raw current-data の local backup、retry を提供します。

## 主要変更

- raw storage port / Chrome adapter / canonical fingerprint / strict repository
- pure legacy analyzer と coordinated preflight service/runtime
- nested URL、category metadata、`domainCategorySettings` の schema・duplicate・dangling relation regression
- blocked/stale React notice、controller、Storybook story、screenshot
- persistence architecture / writer inventory / bootstrap policy の更新

## Acceptance criteria 対応

- raw read-only boundary: forgiving getter・repair・write-back・migration flag 更新なし
- data classification: missing、empty、read/partial failure、invalid inner schema を typed に区別
- identity/relation: exact URL identity、duplicate/collision、nested mismatch、dangling metadata、parent/mapping/order conflict を silent merge/drop せず block
- durability/capacity: versioned deterministic fingerprint、1 MiB minimum + 20% reserve、analysis 後 recheck、明示的 stale
- writer coordination: #711 inventory の normal/implicit/background writer と #727 barrier を共有
- recovery/privacy: raw URL/title/notes/AI/attachments/fingerprint を診断へ含めず、backup は local download のみ、external telemetry なし
- rollout: preflight 先行 release。#728 actual migration と #729 cutover/cleanup は非目標のまま維持

fresh-context Harness Evaluator は changes-requested の semantic edge cases を2回検出し、RED/GREEN修正後の最終評価で **21/21 covered / findings 0 / approved** を確認しました。

## 検証

- `bun run test:coverage`: 381 files、4,346 passed、1 skipped
  - Statements 96.53%、Branches 91.38%、Functions 97.51%、Lines 96.51%
- `bun run quality:check`: pass
- `npx -y react-doctor@latest . --verbose --scope changed --base origin/develop`: 100/100、issues 0
- `bun run release:check`: pass
  - Chrome MV3 / Firefox MV2 production build
  - production logging / app version / network policy verifier
- Harness Evaluator: approved、21/21 covered、findings 0

## Regression risk

- preflight は source data を変更せず、control-plane record 以外へ書きません。
- coordination unavailable、read/fingerprint/capacity failure、source change、ambiguous legacy relation はすべて fail closed です。
- permission、manifest、dependency、external network call の変更はありません。

## UI

Migration preflight blocked notice

Closes #738

## Summary by CodeRabbit

* **新機能**
  * データ移行前の自動診断（移行前チェック）を追加し、状態に応じてアプリ上で警告表示されます。
  * 問題がある場合は、警告とともに「診断情報のコピー」「現在データのバックアップ」「再試行」を選べます。
* **改善**
  * 診断は個人データやURLなどの内容を含めない安全な概要のみを扱い、状態（健全／要確認／停止など）を保持して不必要な再診断を抑えます。
* **ドキュメント**
  * 移行前診断の状態遷移、容量条件、回復手順を整理しました。
* **テスト**
  * 移行前チェックのUI・サービス挙動と診断生成を検証するテストを追加しました。

### CODE REVIEW

File: docs/architecture/current-storage-writer-inventory.md [hunk 1]
```diff
@@ -142,9 +143,11 @@ by omitting an argument.
...
 | background tab created handler | background `tabs.onCreated` handler delegates to `src/lib/background/url-storage.ts`                                |
 
-The direct control-state repository is the sole raw local-storage exception
-inside migration/control plumbing; it must not call the gated facade
-recursively. Settings, UI theme, and release
+The direct control-state repository and #738's raw reader/preflight repository
+are the raw local-storage exceptions inside migration/control plumbing; they
+must not call the gated facade recursively. Both are invoked only while the
+shared #727 coordination adapter holds the exclusive lock. Settings, UI theme,
+and release
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply docs/architecture/current-storage-writer-inventory.md]: 採用しました。82dded7b で文書を実装契約に合わせ、exclusive lock保証をraw source snapshot/recheckとpreflight record writeに限定し、control/preflight record readはlock外になり得ることを明記しました。検証: format、architecture check通過。

File: docs/architecture/persistence-model-v2.md [hunk 2]
```diff
@@ -730,6 +730,18 @@ before any migration write, and persists `PERSISTENCE_PREFLIGHT_STALE` on a
 mismatch. #738 remains responsible for producing the read-only preflight
 fingerprint and invalidating its result after normal source writes.
 
+#738 stores that approval separately as `tabbin:migrationPreflight:v1`. The
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply docs/architecture/persistence-model-v2.md]: 採用しました。82dded7b で行頭の #738 を Issue `#738` に変更し、Markdown見出しとして解釈されない形にしました。検証: format/check通過。

File: docs/plans/2026-07-22-issue-738-migration-preflight-design.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+The pure analyzer uses `exact-url-v1`, reports source-only relation and identity
+conflicts before invoking the Issue #712 target checker, and never silently
+merges or drops a record. Missing timestamp provenance is reported; current
+time is not substituted for historical data.
+
+The copied diagnostic contains only fixed issue codes, entity counts, collision
+count, approximate bytes, capacity status, and version identifiers. It excludes
+URLs, titles, notes, keywords, prompts, conversation/attachment content, raw
+errors, and raw fingerprints. No external telemetry is added.
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply docs/plans/2026-07-22-issue-738-migration-preflight-design.md:69]: 採用しました。82dded7b でcopied diagnosticの契約から、実際には保存・コピーされない approximate bytes を削除しました。検証: diagnostic型・decoderとの一致を確認、format/check通過。

File: src/app/composition/MigrationPreflightNotice.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    >
+      <div>
+        <h2 className='font-semibold'>
+          {t('options.migrationPreflight.title', '移行前チェックが必要です')}
+        </h2>
+        <p className='text-sm text-muted-foreground'>
+          {t(
+            'options.migrationPreflight.description',
+            '現在のデータは変更されていません。',
+          )}
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply src/app/composition/MigrationPreflightNotice.tsx]: 採用しました。82dded7b で stale は「現在のデータは変更されています。」、blocked は従来文言に分岐し、個別regression testを追加しました。同時に通知文のcontrastを4.5:1以上に修正。検証: DOM 5件、axe E2E 2件通過。

File: src/contexts/saved-tabs/application/services/LegacyMigrationPreflightAnalyzerService.ts [hunk 5]
```diff
@@ -0,0 +1,1293 @@
...
+  try {
+    approximateSourceBytes = measureSerializedBytes(createSourcePayload(source))
+  } catch {
+    addIssue(state, 'NON_JSON_SAFE_VALUE', 'error')
+  }
+  const targetSerializedBytes = measureSerializedBytes({
+    analyticsViews,
+    conversations,
+    snapshot,
+  })
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply src/contexts/saved-tabs/application/services/LegacyMigrationPreflightAnalyzerService.ts]: 採用しました。82dded7b で targetSerializedBytes も try/catchし、非JSON-safe値は NON_JSON_SAFE_VALUE を積んで0 bytesとしてfail-closedに返すよう変更しました。BigInt入力のregression testを追加し、node tests通過。

File: src/contexts/saved-tabs/application/services/MigrationPreflightService.ts [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+          return toPublicStatus(record)
+        }
+        return this.saveStale(record)
+      })
+    } catch {
+      return this.options.coordination.runExclusive(async () =>
+        this.saveStale(record),
+      )
+    }
+  }
```
[the remark this answers is not in the document]
Comment (haruto-yamada1) [reply src/contexts/saved-tabs/application/services/MigrationPreflightService.ts:115]: 採用しました。82dded7b で readStatus() の再読取例外を stale に丸めず saveReadFailure() へ流し、typed blocked recordを保存するよう修正しました。raw read failureのregression testを追加し、MigrationPreflightService tests通過。
